src/node/plugin.ts

```diff
@@ -1,92 +1,189 @@
+interface Plugin extends pluginapi.Plugin {
+  /**
+   * These fields are populated from the plugin's package.json.
+   */
+  name: string
+  version: string
+  description: string
 
-export type Activate = (app: Express, args: Args) => void
-
-/**
- * Plugins must implement this interface.
- */
-export interface Plugin {
-  activate: Activate
+  /**
+   * path to the node module on the disk.
+   */
+  modulePath: string
 }
 
-/**
- * Intercept imports so we can inject code-server when the plugin tries to
- * import it.
- */
-const originalLoad = require("module")._load
-// eslint-disable-next-line @typescript-eslint/no-explicit-any
-require("module")._load = function (request: string, parent: object, isMain: boolean): any {
-  return originalLoad.apply(this, [request.replace(/^code-server/, path.resolve(__dirname, "../..")), parent, isMain])
+interface Application extends pluginapi.Application {
+  plugin: Plugin
 }
 
 /**
- * Load a plugin and run its activation function.
+ * PluginAPI implements the plugin API described in typings/plugin.d.ts
+ * Please see that file for details.
  */
-const loadPlugin = async (pluginPath: string, app: Express, args: Args): Promise<void> => {
-  try {
-    const plugin: Plugin = require(pluginPath)
-    plugin.activate(app, args)
-
-    const packageJson = require(path.join(pluginPath, "package.json"))
-    logger.debug(
-      "Loaded plugin",
-      field("name", packageJson.name || path.basename(pluginPath)),
-      field("path", pluginPath),
-      field("version", packageJson.version || "n/a"),
-    )
-  } catch (error) {
-    logger.error(error.message)
+export class PluginAPI {
+  private readonly plugins = new Array<Plugin>()
+  private readonly logger: Logger
+
+  public constructor(
+    logger: Logger,
+    /**
+     * These correspond to $CS_PLUGIN_PATH and $CS_PLUGIN respectively.
+     */
+    private readonly csPlugin = "",
+    private readonly csPluginPath = `${path.join(util.paths.data, "plugins")}:/usr/share/code-server/plugins`,
+  ) {
+    this.logger = logger.named("pluginapi")
   }
-}
 
-/**
- * Load all plugins in the specified directory.
- */
-const _loadPlugins = async (pluginDir: string, app: Express, args: Args): Promise<void> => {
-  try {
-    const files = await util.promisify(fs.readdir)(pluginDir, {
-      withFileTypes: true,
-    })
-    await Promise.all(files.map((file) => loadPlugin(path.join(pluginDir, file.name), app, args)))
-  } catch (error) {
-    if (error.code !== "ENOENT") {
-      logger.warn(error.message)
+  /**
+   * applications grabs the full list of applications from
+   * all loaded plugins.
+   */
+  public async applications(): Promise<Application[]> {
+    const apps = new Array<Application>()
+    for (const p of this.plugins) {
+      const pluginApps = await p.applications()
+
+      // Add plugin key to each app.
+      apps.push(
+        ...pluginApps.map((app) => {
+          return { ...app, plugin: p }
```

Comment:
I'm not sure this really matters but it feels a bit strange to me that the whole plugin is here (even though it's not on the types). My thinking is that it might cause some unexpected behavior (for example like how you had to `stringify` in the tests).

Reply:
Fair enough! I'll be explicit here.